### PR TITLE
Browser pages

### PR DESCRIPTION
**Pull request by JoelEinbinder**
This adds new events `pageopened` and `pageclosed` to browser.
It also adds `browser.pages()`.
#386 #443

### CODE REVIEW

File: lib/Browser.js [hunk 1]
```diff
@@ -16,18 +16,69 @@
...
+    const browser =  new Browser(connection, ignoreHTTPSErrors, closeCallback);
+    await connection.send('Target.setDiscoverTargets', {discover: true});
+    return browser;
+  }
+
+  /**
+   * @param {{targetInfo: {targetId: string}}} param0
+   */
+  async _targetCreated({targetInfo}) {
+    if (targetInfo.type !== 'page' && targetInfo.type !== 'other')
```
Comment (JoelEinbinder) [inline lib/Browser.js]: The newtab pages come in as 'other'. I'm not sure whether this is intentional or not, but it makes them hard to work with.

**Comment by vsemozhetbyt** [comment]
Sorry, it seems the version in the heading of [this doc](https://github.com/GoogleChrome/puppeteer/blob/v0.10.0/docs/api.md) needs to be updated into 0.10.0.

**Comment by twalpole** [comment]
I believe these tests fail in non-headless mode, as soon as `window.open` is run, with "Unhandled promise rejection (rejection id: 1): Error: Protocol error (Network.enable): Target closed."  I think it's due to Chrome shifting tab focus, but not really sure.

**Comment by twalpole** [comment]
Additionally, it appears the main frame is not attached to the page yet when `pageopened` is  fired, limiting the ability to do things like  `injectFile`, etc.  Is this intended?

File: lib/Browser.js (hunk 1 above)
Comment (dgozman) [reply lib/Browser.js]: Please file an upstream bug instead of this.

**Comment by duckywang1** [comment]
This PR is awesome!   
I am eager to get something like `browser.pages()`. So I did some tests with [Branch: browser_pages](https://github.com/JoelEinbinder/puppeteer/tree/browser_pages).

What I did:
1. try it with `headless: true`.  
The results were great. I could get pages which comes from `page.goto()` and which created from `a[target="_blank"]` 

2. try it with `headless: false`.  
Error came out. `nodejs.unhandledRejectionError: Protocol error (Network.enable): Target closed`. Just like what @twalpole said.

**Comment by twalpole** [comment]
@duckywang1 When running headless, were you able to actually do anything/interact with the pages created from `a[target="_blank"]` ?

**Comment by duckywang1** [comment]
Apologies. I have no experience of unit test.
Yesterday I didn't try this feature seriously. Today I write an example of my business logic with `browser.pages();` and `headless browser`. It worked!
I can interact with page created from `a[target="_blank"]`. @twalpole 

Now I hope this great feature would be merged, althought it only take effect in `headless browser` temporarily.

Here is my example:  
I got two pages.
1. detail page.
This page provide an HTMLElememt `a[target="_blank"]` and link to download page
2. download page.
This page will provide a resource URL and download it in 2-10 second automatically.

Here is my example code(fake code, but it did work):
```
const browser = initHeadlessBrowser(); // create a Headless browser
const detailPage = browser.newPage();
detailPage.goto(url, '{ waitUntil: 'networkidle' }');
const btn = page.$('.download');
btn.click({ delay: 100 }); // create a download page but we can fetch them with browser.pages();

const allPages = browser.pages();
for (let i = 1; i < allPages.length; i += 1) {
  // allPages[0] is a blank page. We can't get any valid info. If interested, you can open a headless: false browser and see the GUI of Chrominum.
  // compare something with your download page
  // here I confirm it with page.url();
  if (condition) {
    const downloadPage = allPages[i];
    await downloadPage.setRequestInterceptionEnabled(true);
    downloadPage.on('request', request => {
      if (request.url.indexOf('.png') >= 0) {
        // here i use node_modules`download`
        download(request.url).then(streamData => {
          // get streamData and do something fun
        });
      } else {
        request.continue();
      }
    });
    await downloadPage.waitFor(30000); // waiting for loading views
    await downloadPage.close(); // close download page
  }
}

```
By the way, I think it seems a little strange of `how to use the downloadPage` or `lifeCycle of downloadPage`. But it worked and suprised me.

**Comment by twalpole** [comment]
@duckywang1 Ah -- I see you're using `pages` not the 'pageopened' and 'pageclosed' events

**Comment by nwhitmont** [comment]
Looking forward to this!

**Comment by avimar** [comment]
Can someone explain this?
>The newtab pages come in as 'other'

... it seems unnecessary to launch a new browser and _always_ having to open a new tab. Why can't I just use the one that's already open?

**Comment by dmportella** [comment]
Any more work being done on this?

**Comment by nwhitmont** [comment]
@JoelEinbinder What's the latest update on Browser Pages?

**Comment by schelkun** [comment]
I'm waiting for this PR will be accepted. Any problems with this PR?

**Comment by dmportella** [comment]
i am trying to use your changes to capture a popup but it doesnt seem to work.. there are no errors but no pagecreated event is fired

I could share the snippet - could be that i am doing something wrong

**Comment by JoelEinbinder** [comment]
The Chromium can be rolled to r501642 to fix targets appearing as `"other"`.

@dmportella Sharing the snippet would be very useful to me! Thanks.

**Comment by nwhitmont** [comment]
@dgozman @JoelEinbinder @vsemozhetbyt 

Hey guys, can you help us understand what is preventing this from being accepted/merged?

**Comment by aslushnikov** [comment]
@nwhitmont this PR doesn't work well enough. One of the issues is that Puppeteer's FrameManager doesn't know how to attach to a non-empty page.

We're working on it; we'll merge this PR as all the blockers are eliminated.

**Review by ghost — APPROVED** [review]

**Comment by JoelEinbinder** [comment]
There is still a race somewhere when running in headful mode. `Target.attachToTarget` returns before the target is usable, and so the first message to it sometimes gets rejected with a target closed error.

**Comment by nwhitmont** [comment]
Thanks for the updates & keep up the great work!

**Comment by zxsoft** [comment]
Hope this PR can be approved in the next release. I need it so badly.

I hold a bunch of page handlers (because of no browser.pages() ), and when a page is crashed, i don't know. So my page handlers are not reliable. And I cannot make it better. because I don't know wether a page is available or not.

**Comment by trafostream** [comment]
How to setRequestInterceptionEnabled on new popup pages (target="_blank") and attach event on request. For example, I need block the download popup pages by url.
Thank you.
Sorry for my english.

I would like something like this:

```
browser.on('pageopened',  pageOpened => {
        pageOpened.setRequestInterceptionEnabled(true);
        pageOpened.on('request', request => {
              request.abort();
        });
});
```

File: lib/Browser.js [hunk 2]
```diff
@@ -30,6 +30,55 @@ class Browser extends EventEmitter {
...
+  }
+
+  /**
+   * @param {!Connection} connection
+   * @param {boolean} ignoreHTTPSErrors
+   * @param {function()=} closeCallback
+   */
+  static async create(connection, ignoreHTTPSErrors, closeCallback) {
+    const browser =  new Browser(connection, ignoreHTTPSErrors, closeCallback);
+    await connection.send('Target.setDiscoverTargets', {discover: true});
```
Comment (pavelfeldman) [inline lib/Browser.js:48]: You should call `Target.getTargets` to read present state and also you should modify it do the same thing `Target.setDiscoverTargets` does. That way you don't have a race between these calls and target creation.

File: lib/Browser.js [hunk 3]
```diff
@@ -30,6 +30,55 @@ class Browser extends EventEmitter {
...
+  }
+
+  /**
+   * @param {string} targetId
+   * @return {Page}
+   */
+  _pageForTargetId(targetId) {
+    if (this._pages.has(targetId))
+      return this._pages.get(targetId);
+    const pagePromise = this._connection.createSession(targetId).then(client => Page.create(client, this._ignoreHTTPSErrors, this._screenshotTaskQueue));
```
Comment (pavelfeldman) [inline lib/Browser.js]: I don't think proactively attaching to all the targets is necessary. I'd rather expose the API as it is in the protocol - you can list the targets and you can materialize them as pages...

File: lib/Page.js [hunk 4]
```diff
@@ -44,9 +44,9 @@ class Page extends EventEmitter {
     if (ignoreHTTPSErrors)
       await client.send('Security.setOverrideCertificateErrors', {override: true});
     const page = new Page(client, ignoreHTTPSErrors, screenshotTaskQueue);
-    await page.goto('about:blank');
     // Initialize default page size.
     await page.setViewport({width: 800, height: 600});
+    await page._frameManager.fetchResourceTree();
```
Comment (pavelfeldman) [inline lib/Page.js]: You are fetching resource tree after initializing Runtime - that way you won't get execution contexts mapped to the frames.

File: lib/Browser.js (hunk 2 above)
Comment (dgozman) [reply lib/Browser.js:48]: Please don't :) Target.setDiscoverTargets should notify about all existing ones through Target.targetCreated. There is no need to use Target.getTargets at all.

Comment (pavelfeldman) [reply lib/Browser.js:48]: Well, if you look at [TargetHandler::SetDiscoverTargets](https://cs.chromium.org/chromium/src/content/browser/devtools/protocol/target_handler.cc?q=TargetHandler::SetD&sq=package:chromium&l=187), you would be certain it does not. Except for I now see that your observer has side effects. But that is just mean, you know, calling it observer and making it change the model!

File: docs/api.md [hunk 5]
```diff
@@ -223,6 +242,12 @@ Closes browser with all the pages (if any were opened). The browser object itsel
 #### browser.newPage()
 - returns: <[Promise]<[Page]>> Promise which resolves to a new [Page] object.
 
+#### browser.pages()
+- returns: <[Promise]<[Array]<[Page]>>> Promise which resolves to an array of all open pages.
+
+#### browser.targets()
+- returns: <[Promise]<[Array]<[Page]>>> Promise which resolves to an array of all open pages.
```
Comment (aslushnikov) [inline docs/api.md]: to an array of all open targets

File: docs/api.md [hunk 6]
```diff
@@ -1439,6 +1464,24 @@ Contains the status code of the response (e.g., 200 for a success).
...
 - <[string]>
 
+### class: Target
+
+#### target.page()
+- returns: <[Promise]<[Page]>>
+
+If the target is not of type `"page"`, returns `null`.
+
+#### target.title()
```
Comment (aslushnikov) [inline docs/api.md]: let's drop it?

File: lib/Browser.js [hunk 7]
```diff
@@ -81,3 +147,61 @@ class TaskQueue {
...
+  }
+
+  /**
+   * @return {Promise<string>}
+   */
+  async url() {
+    return (await this._info()).url;
+  }
+
+  type() {
```
Comment (aslushnikov) [inline lib/Browser.js]: nit: jsdoc

File: lib/Browser.js [hunk 8]
```diff
@@ -81,3 +147,61 @@ class TaskQueue {
...
+    return this._type;
+  }
+}
+helper.tracePublicAPI(Target);
+
+/**
+ * @typedef {Object} Target.targetInfo
```
Comment (aslushnikov) [inline lib/Browser.js]: this is never used

File: lib/Page.js [hunk 9]
```diff
@@ -36,6 +36,9 @@ class Page extends EventEmitter {
    * @return {!Promise<!Page>}
    */
   static async create(client, ignoreHTTPSErrors, appMode, screenshotTaskQueue) {
+    const page = new Page(client, ignoreHTTPSErrors, screenshotTaskQueue);
+    page._frameManager.fetchResourceTree();
```
Comment (aslushnikov) [inline lib/Page.js]: don't you want to await it?

File: lib/Page.js [hunk 10]
```diff
@@ -44,8 +47,6 @@ class Page extends EventEmitter {
     ]);
     if (ignoreHTTPSErrors)
       await client.send('Security.setOverrideCertificateErrors', {override: true});
-    const page = new Page(client, ignoreHTTPSErrors, screenshotTaskQueue);
```
Comment (aslushnikov) [inline lib/Page.js:49]: why this change?

File: lib/Browser.js [hunk 11]
```diff
@@ -31,6 +31,50 @@ class Browser extends EventEmitter {
...
+      return;
+    this._targets.delete(targetId);
+    this.emit(Browser.Events.TargetDestroyed, target);
+  }
+
+  /**
+   * @param {{targetId: string, type: string}} targetInfo
+   * @return {Target}
+   */
+  _targetForInfo(targetInfo) {
```
Comment (aslushnikov) [inline lib/Browser.js]: nit: this kind of methods are usually prefixed with `_ensure`: `_ensureTargetForInfo`

File: lib/Browser.js [hunk 12]
```diff
@@ -45,8 +89,24 @@ class Browser extends EventEmitter {
...
-    const client = await this._connection.createSession(targetId);
-    return await Page.create(client, this._ignoreHTTPSErrors, this._appMode, this._screenshotTaskQueue);
+    const {targetInfo} = await this._connection.send('Target.getTargetInfo', {targetId});
+    const page = await this._targetForInfo(targetInfo).page();
+    await page.goto('about:blank');
+    return page;
+  }
+
+  /**
+   * @return {Array<Target>}
```
Comment (aslushnikov) [inline lib/Browser.js]: For the sake of consistency, please either:
- use `'!'` for non-nullable types in this patch
- remove all exclamation marks throughout the code first

File: lib/Browser.js [hunk 13]
```diff
@@ -45,8 +89,24 @@ class Browser extends EventEmitter {
...
+   */
+  targets() {
+    return Array.from(this._targets.values());
+  }
+
+  /**
+   * @return {Array<Pages>}
```
Comment (aslushnikov) [inline lib/Browser.js]: Promise type declaration is missing

I also like `async` methods since they make it really obvious that the method should be awaited. Let's make `pages` async!

File: lib/Browser.js [hunk 14]
```diff
@@ -45,8 +89,24 @@ class Browser extends EventEmitter {
...
+   */
+  pages() {
+    return Promise.all(this.targets().map(target => target.page()).filter(page => page));
```
Comment (aslushnikov) [inline lib/Browser.js]: nit: `page => !!page`

File: lib/FrameManager.js [hunk 15]
```diff
@@ -40,6 +40,20 @@ class FrameManager extends EventEmitter {
...
+    const handleTree = tree => {
+      this._onFrameNavigated(tree.frame);
+      if (!tree.childFrames)
+        return;
+
+      for (const child of tree.childFrames)
+        handleTree(child);
+    };
+
+    const { frameTree } = await this._client.send('Page.getResourceTree');
```
Comment (aslushnikov) [inline lib/FrameManager.js]: frameTree should be fetched in `page.create` static method and passed in FrameManager constructor as an argument

File: lib/Browser.js [hunk 16]
```diff
@@ -81,3 +147,61 @@ class TaskQueue {
...
+    this._targetId = targetId;
+    this._type = type;
+    /** @type {Promise<Page>} */
+    this._pagePromise = null;
+  }
+
+  /**
+   * @return {Promise<{targetId: string, type: string, title: string, url: string, attached: boolean, type: string}>}
+   */
+  async _info() {
```
Comment (aslushnikov) [inline lib/Browser.js]: you should create target with targetInfo and make this sync.
You'll also need to listen to `targetInfoChanged` event.

**Review by aslushnikov — COMMENTED** [review]
Glad to see this is happenning!

File: lib/Page.js (hunk 10 above)
Comment (JoelEinbinder) [reply lib/Page.js:49]: The FrameManager needs to be listening for execution contexts before Runtime.enable is called.

File: lib/Page.js [hunk 17]
```diff
@@ -36,34 +36,41 @@ class Page extends EventEmitter {
...
-    await page.goto('about:blank');
     // Initialize default page size.
     if (!appMode)
       await page.setViewport({width: 800, height: 600});
+
+    // The page might still be initializing, if so wait for the first navigation
```
Comment (aslushnikov) [inline lib/Page.js]: @JoelEinbinder could you please check with @dgozman? this looks like something that should not be happenning

File: lib/Browser.js [hunk 18]
```diff
@@ -81,3 +155,51 @@ class TaskQueue {
     return result;
   }
 }
+
+class Target {
+  /**
+   * @param {Browser} browser
+   * @param {Target.TargetInfo} targetInfo
```
Comment (aslushnikov) [inline lib/Browser.js]: nit: '!'

File: lib/Browser.js [hunk 19]
```diff
@@ -31,6 +31,58 @@ class Browser extends EventEmitter {
...
+      return;
+    this._targets.delete(targetId);
+    this.emit(Browser.Events.TargetDestroyed, target);
+  }
+
+  /**
+   * @param {{targetInfo: Target.TargetInfo}} param0
+   */
+  async _targetInfoChanged({targetInfo}) {
+    this._ensureTargetForInfo(targetInfo)._targetInfo = targetInfo;
```
Comment (aslushnikov) [inline lib/Browser.js]: do we want to emit 'targetChanged' here?

File: lib/Page.js (hunk 17 above)
Comment (JoelEinbinder) [reply lib/Page.js]: "It is good enough" - @dgozman

Comment (aslushnikov) [reply lib/Page.js]: I don't understand why we do this here. Doesn't it make more sense in target reporting code then?

File: lib/Browser.js (hunk 19 above)
Comment (JoelEinbinder) [reply lib/Browser.js]: Let's add in a later patch.

File: lib/Page.js (hunk 17 above)
Comment (JoelEinbinder) [reply lib/Page.js]: It is accurately reporting whats going on in the browser. it just is inconvenient for users of `Page` so we wait for first navigation.

File: test/test.js [hunk 20]
```diff
@@ -2472,6 +2472,64 @@ describe('Page', function() {
...
+      expect(await closePagePromise).toBe(otherPage);
+
+      allPages = await Promise.all(browser.targets().map(target => target.page()));
+      expect(allPages).toContain(page);
+      expect(allPages).not.toContain(otherPage);
+    }));
+    it('should report when a service worker is created and destroyed', SX(async function() {
+      await page.goto(EMPTY_PAGE);
+      const createdTarget = new Promise(fulfill => browser.once('targetcreated', target => fulfill(target)));
+      const registration = await page.evaluateHandle(() => navigator.serviceWorker.register('sw.js'));
```
Comment (aslushnikov) [inline test/test.js:2797]: what is 'sw.js'? Should you add the file to the patch?

File: test/test.js [hunk 21]
```diff
@@ -2472,6 +2472,64 @@ describe('Page', function() {
 
     }));
   });
+
+  describe('Target', function() {
+    it('Browser.targets should return all of the targets', SX(async function() {
```
Comment (aslushnikov) [inline test/test.js:2751]: this and the following test seem to be doing the same thing

File: test/test.js [hunk 22]
```diff
@@ -2472,6 +2472,64 @@ describe('Page', function() {
...
+      // The pages will be the testing page and the original newtab page
+      const targets = browser.targets();
+      expect(targets.some(target => target.type() === 'page' &&
+        target.url() === 'about:blank')).toBeTruthy('Missing blank page');
+      expect(targets.some(target => target.type() === 'other' &&
+        target.url() === '')).toBeTruthy('Missing browser target');
```
Comment (aslushnikov) [inline test/test.js:2757]: what's the argument in `toBeTruthy`?

File: lib/Page.js (hunk 17 above)
Comment (aslushnikov) [reply lib/Page.js]: is there any test that would help me understand what's going on?

File: test/test.js (hunk 20 above)
Comment (JoelEinbinder) [reply test/test.js:2797]: I did, it is just empty.

File: test/test.js (hunk 22 above)
Comment (JoelEinbinder) [reply test/test.js:2757]: It is the error message when the condition fails.

**Comment by aslushnikov** [comment]
As discussed offline, these are the last bits remaining:
- [x] move target initialization "hack" to target creation
- [x] report targetchanged event so that URL changes are observable

**Comment by aslushnikov** [comment]
For the record: this is waiting for the [upstream change](https://chromium-review.googlesource.com/c/chromium/src/+/711215) to land.

**Comment by schelkun** [comment]
Is it working with latest code from master?

**Comment by JoelEinbinder** [comment]
Should be good now that #1019 landed.

File: lib/Browser.js [hunk 23]
```diff
@@ -31,6 +31,64 @@ class Browser extends EventEmitter {
...
+
+  /**
+   * @param {{targetInfo: Target.TargetInfo}} event
+   */
+  _targetInfoChanged(event) {
+    const target = this._ensureTargetForInfo(event.targetInfo);
+    const previousURL = target._targetInfo.url;
+    target._targetInfo = event.targetInfo;
+    if (previousURL !== event.targetInfo.url)
+      this.emit(Browser.Events.TargetURLChanged, target);
```
Comment (aslushnikov) [inline lib/Browser.js]: let's call this `targetchanged`: if we're ever to expose anything else on target (e.g. target's title), it would cover the use case.

File: lib/Browser.js [hunk 24]
```diff
@@ -31,6 +31,64 @@ class Browser extends EventEmitter {
...
+
+  /**
+   * @param {!Target.TargetInfo} targetInfo
+   * @return {!Target}
+   */
+  _ensureTargetForInfo(targetInfo) {
+    if (!this._targets.has(targetInfo.targetId)) {
+      const target = new Target(this, targetInfo);
+      this._targets.set(targetInfo.targetId, target);
+      this.emit(Browser.Events.TargetCreated, target);
```
Comment (aslushnikov) [inline lib/Browser.js]: let's not emit in-between: exception in one of user's listeners would corrupt our state.

File: lib/Browser.js [hunk 25]
```diff
@@ -80,4 +161,59 @@ class TaskQueue {
...
+
+class Target {
+  /**
+   * @param {Browser} browser
+   * @param {!Target.TargetInfo} targetInfo
+   */
+  constructor(browser, targetInfo) {
+    this._browser = browser;
+    this._targetInfo = targetInfo;
+    /** @type {!Promise<!Page>} */
```
Comment (aslushnikov) [inline lib/Browser.js]: nit: promise is nullable in this case

File: lib/Browser.js [hunk 26]
```diff
@@ -80,4 +161,59 @@ class TaskQueue {
...
+    this._pagePromise = null;
+  }
+
+  /**
+   * @return {?Promise<!Page>}
```
Comment (aslushnikov) [inline lib/Browser.js]: let's return `!Promise<?Page>` instead of `?Promise<!Page>`

File: lib/Browser.js [hunk 27]
```diff
@@ -80,4 +161,59 @@ class TaskQueue {
...
+  url() {
+    return this._targetInfo.url;
+  }
+
+  /**
+   * @return {"page"|"service_worker"|"other"}
+   */
+  type() {
+    const type = this._targetInfo.type;
+    switch (type) {
```
Comment (aslushnikov) [inline lib/Browser.js]: nit: i'd rather have this as an if.
```js
const type = this._targetInfo.type;
if (type === 'page' || type === 'service_worker')
  return type;
return 'other';
```

**Comment by schelkun** [comment]
So now I should use
```
  browser.on('targetcreated',  pageOpened => {
      console.log('1');
  });
```

Right?

File: lib/Browser.js (hunk 23 above)
Comment (JoelEinbinder) [reply lib/Browser.js]: Done.

File: lib/Browser.js (hunk 24 above)
Comment (JoelEinbinder) [reply lib/Browser.js]: Done.

File: lib/Browser.js (hunk 25 above)
Comment (JoelEinbinder) [reply lib/Browser.js]: Done.

File: lib/Browser.js (hunk 26 above)
Comment (JoelEinbinder) [reply lib/Browser.js]: Done.

File: lib/Browser.js (hunk 27 above)
Comment (JoelEinbinder) [reply lib/Browser.js]: Done.

File: lib/Browser.js [hunk 28]
```diff
@@ -31,6 +31,61 @@ class Browser extends EventEmitter {
...
+    await connection.send('Target.setDiscoverTargets', {discover: true});
+    return browser;
+  }
+
+  /**
+   * @param {{targetInfo: Target.TargetInfo}} event
+   */
+  async _targetCreated(event) {
+    const target = this._ensureTargetForInfo(event.targetInfo);
+    await target._initializedPromise;
```
Comment (aslushnikov) [inline lib/Browser.js]: this will never resolve if target was killed before initialization.

File: lib/Browser.js [hunk 29]
```diff
@@ -45,8 +111,25 @@ class Browser extends EventEmitter {
    */
   async newPage() {
     const {targetId} = await this._connection.send('Target.createTarget', {url: 'about:blank'});
-    const client = await this._connection.createSession(targetId);
-    return await Page.create(client, this._ignoreHTTPSErrors, this._appMode, this._screenshotTaskQueue);
+    const target = await this._targetForId(targetId);
```
Comment (aslushnikov) [inline lib/Browser.js]: Let's inline it

```js
const target = this._targets.get(targetId);
console.assert(target, 'Failed to create target for page');
await target._initializedPromise;
```

File: lib/Browser.js [hunk 30]
```diff
@@ -80,4 +170,77 @@ class TaskQueue {
...
+  /**
+   * @param {!Browser} browser
+   * @param {!Target.TargetInfo} targetInfo
+   */
+  constructor(browser, targetInfo) {
+    this._browser = browser;
+    this._targetInfo = targetInfo;
+    /** @type {?Promise<!Page>} */
+    this._pagePromise = null;
+    this._wasInitialized = false;
```
Comment (aslushnikov) [inline lib/Browser.js]:
```js
this._isInitialized = this._targetInfo.type !== 'page' || this._targetInfo.url !== '');

```

File: lib/Browser.js [hunk 31]
```diff
@@ -80,4 +170,77 @@ class TaskQueue {
...
+    this._initializedPromise = new Promise(fulfill => this._initialized = fulfill);
+
+    this._targetInfoChanged(targetInfo);
```
Comment (aslushnikov) [inline lib/Browser.js]: this would not be needed

File: lib/FrameManager.js [hunk 32]
```diff
@@ -40,6 +41,20 @@ class FrameManager extends EventEmitter {
...
     this._client.on('Runtime.executionContextCreated', event => this._onExecutionContextCreated(event.context));
+
+    this._handleFrameTree(frameTree);
+  }
+
+  /**
+   * @param {{frame: Object, childFrames: ?Array}} frameTree
+   */
+  _handleFrameTree(frameTree) {
+    this._onFrameNavigated(frameTree.frame);
```
Comment (aslushnikov) [inline lib/FrameManager.js]: this shouldn't work? Lets's attach to somthing with nested frames

File: lib/Browser.js [hunk 33]
```diff
@@ -80,4 +170,77 @@ class TaskQueue {
...
+
+  /**
+   * @param {!Target.TargetInfo} targetInfo
+   */
+  _targetInfoChanged(targetInfo) {
+    const previousURL = this._targetInfo.url;
+    this._targetInfo = targetInfo;
+
+    if (!this._wasInitialized && (this._targetInfo.type !== 'page' || this._targetInfo.url !== '')) {
+      this._wasInitialized = true;
```
Comment (aslushnikov) [inline lib/Browser.js]: s/_wasInitialized/_isInitialized
s/_initialzed/_initializeCallback

File: lib/Browser.js [hunk 34]
```diff
@@ -31,6 +31,61 @@ class Browser extends EventEmitter {
...
+    this._targets.delete(event.targetId);
+    if (await target._initializedPromise)
+      this.emit(Browser.Events.TargetDestroyed, target);
+  }
+
+  /**
+   * @param {{targetInfo: !Target.TargetInfo}} event
+   */
+  _targetInfoChanged(event) {
+    const target = this._ensureTargetForInfo(event.targetInfo);
```
Comment (aslushnikov) [inline lib/Browser.js]: assert target here; never create

File: lib/Browser.js [hunk 35]
```diff
@@ -31,6 +31,61 @@ class Browser extends EventEmitter {
...
+    const browser =  new Browser(connection, ignoreHTTPSErrors, closeCallback);
+    await connection.send('Target.setDiscoverTargets', {discover: true});
+    return browser;
+  }
+
+  /**
+   * @param {{targetInfo: !Target.TargetInfo}} event
+   */
+  async _targetCreated(event) {
+    const target = this._ensureTargetForInfo(event.targetInfo);
```
Comment (aslushnikov) [inline lib/Browser.js]: create target here; it should never exist before

**Review by aslushnikov — APPROVED** [review]

**Comment by ebidel** [comment]
Wooooooot!

**Comment by pirate** [comment]
Thanks everyone who worked on this, it looks great!  Really appreciate this change, it'll make life a lot easier for us when testing popups!

**Comment by vivien7512** [comment]
so this great feature  is available  from which number version?

**Comment by nwhitmont** [comment]
Sweet!

**Comment by BlackCapCoder** [comment]
Thank you for working on this! I am making a vi-like web automation/scraping language, but it has been on hold for some time because this is absolutely essential.

**Comment by manideep1111** [comment]
can you give any code sample of, how to open multiple tabs using puppeteer?

**Comment by JoelEinbinder** [comment]
@manideep1111 you can use [`browser.newPage`](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#browsernewpage) more than once to create more tabs.

**Comment by manideep1111** [comment]
Hi JoelEinbinder Thanks for your reply, I tried with  browser.newPage but my side giving UnhandlePromiseRejection error came .
If possible plea send any example code.

**Comment by manideep1111** [comment]
Thanks @JoelEinbinder for reply, I used browser.newPage my side it's not opening new tab,If you give any code example it's grateful for me.

**Comment by JoelEinbinder** [comment]
@manideep1111 can you file a separate issue? Thanks!

**Comment by manideep1111** [comment]
No separate issue ? If you want see my file I can send to you..Thanks
